draft-ietf-jsonpath-base.md

```diff
@@ -1023,22 +1026,23 @@ logical-or-expr  = logical-and-expr *(S "||" S logical-and-expr)
 logical-and-expr = basic-expr *(S "&&" S basic-expr)  ; conjunction
                                                       ; binds more tightly than disjunction
 
-basic-expr   = exist-expr /
-               paren-expr /
-               relation-expr
-exist-expr   = [neg-op S] path                          ; path existence or non-existence
-path         = rel-path / json-path
-rel-path     = "@" *(S (dot-selector / index-selector))
-paren-expr   = [neg-op S] "(" S boolean-expr S ")"    ; parenthesized expression
-neg-op       = "!"                                    ; not operator
+basic-expr        = exist-expr /
+                    paren-expr /
+                    relation-expr
+exist-expr        = [neg-op S] singular-path          ; path existence or non-existence
```

Comment:
Why does an exist expression need to be singular?

Reply:
Relative pathes always should be singular.
```
During iteration process each array element or object member is
visited and its value -- accessible via symbol @ -- or one of its
descendants -- uniquely defined by a relative path -- is tested
against a boolean expression boolean-expr.
```

```
 rel-path     = "@" *(S (dot-selector / index-selector))
```
They seem to be naturally singular.

Reply:
But what about absolute paths?

Reply:
Does the following example help to put you off non-singular absolute paths in filters?
```
$[?(@.a==$.*)]
```

Reply:
hmm ... this should select nothing, since a singular value can't be equal to a nodelist ... implying RHS resulting not in a single value.

So only singular absolute pathes survive in practice ... so why restrict them artificially in ABNF ?

Reply:
> hmm ... this should select nothing, since a singular value can't be equal to a nodelist ... implying RHS resulting not in a single value.
> 
Note though that JSONPath implementations divide into (1) those that use JavaScript, Python, PHP for expressions, and (2) those that invented their own expression language. Considering only the second group, you'll find that the important ones including Jayway allow expressions starting with `@` to be full JSONPath expressions in their own right, and that the result of evaluating these JSONPath expressions within filter expressions are _not_ result lists, rather, they are single values that are resolved according to well defined rules. What you're describing is different than what is found in prior experience, in prior experience, JSONPath expressions appearing in filter expressions do evaluate to single values.

Reply:
> > hmm ... this should select nothing, since a singular value can't be equal to a nodelist ... implying RHS resulting not in a single value.
> 
> Note though that JSONPath implementations divide into (1) those that use JavaScript, Python, PHP for expressions, and (2) those that invented their own expression language. Considering only the second group, you'll find that the important ones including Jayway allow expressions starting with `@` to be full JSONPath expressions in their own right, and that the result of evaluating these JSONPath expressions within filter expressions are _not_ result lists, rather, they are single values that are resolved according to well defined rules. What you're describing is different than what is found in prior experience, in prior experience, JSONPath expressions appearing in filter expressions do evaluate to single values.

If we automatically converted node sets to arrays in filter expressions:

1. This would break existing implementations, including Jayway, in examples such as [$[?(@.key>42)]](https://cburgmer.github.io/json-path-comparison/results/filter_expression_with_greater_than.html) and [$[?(!@.key)]](https://cburgmer.github.io/json-path-comparison/results/filter_expression_with_negation_and_without_value.html) which expect singular paths to produce at most one _value_. (`@.key` would produce a node set with at most one node and this would convert to an empty array or an array with one value.)
2. Objects could produce non-deterministic arrays, which would be tricky to use in comparisons.

Reply:
> If we automatically converted node sets to arrays in filter expressions:
> 
> 1. This would break existing implementations, including Jayway, in examples such as [$[?(@.key>42)]](https://cburgmer.github.io/json-path-comparison/results/filter_expression_with_greater_than.html) and [$[?(!@.key)]](https://cburgmer.github.io/json-path-comparison/results/filter_expression_with_negation_and_without_value.html) which expect singular paths to produce at most one _value_. (`@.key` would produce a node set with at most one node and this would convert to an empty array or an array with one value.)

I don't know of _any_ implementation that automatically converts "results lists" to arrays inside filter expressions. 

Jayway doesn't produce result lists at all when evaluating a JSONPath expression inside a filter. It evaluates them as a single value according to the following rules: The slice, recursive descent and union selectors, which can evaluate to zero, one, or many items, are wrapped in a Json array. The object wildcard selector is wrapped in an array if the object has two or more members, otherwise a single value is returned. The array wildcard operator always returns an array. The others - root, current node, parent node, identifier, and index selectors - evaluate to a single value if matched, otherwise a null value.

As far as I can tell, the important JSONPath implementations that invent their own filter expressions generally allow expressions beginning with '@' to be first class JSONPath expressions. And they evaluate them as single values within filter expressions, they need to produce single values for comparisons with JSON literals to make sense. So they have rules for that, but they don't all follow the same rules as Jayway. Nonetheless, Jayway is probably the most important and influential of these.

The alternative you find in prior experience are the implementations that use JavaScript or Python or PHP for filter expressions. In these, expressions beginning with `@` aren't JSONPath expressions at all. Here `@` is simply a JavaScript or Python or PHP variable, with supported indexing operations. I believe this experience is what Stefan has in mind.

Reply:
> Jayway doesn't produce result lists at all when evaluating a JSONPath expression inside a filter. It evaluates them as a single value according to the following rules: The slice, recursive descent and union selectors, which can evaluate to zero, one, or many items, are wrapped in a Json array. The object wildcard selector is wrapped in an array if the object has two or more members, otherwise a single value is returned. The array wildcard operator always returns an array. The others - current node, identifier, and index selectors - evaluate to a single value if matched, otherwise a null value.

I find this paragraph confusing.

"The slice, recursive descent and union selectors, which can evaluate to zero, one, or many items, are wrapped in a Json array. " -- I don't know what it means to wrap a selector in a JSON array. I presume it should say the node sets produced by these selectors  are wrapped in JSON arrays, which sounds very much like automatically converting node sets into arrays.

"The object wildcard selector is wrapped in an array if the object has two or more members, otherwise a single value is returned." -- I presume you mean the node set produced by such a selector is wrapped (in a non-deterministic order?) in a JSON array if it has two or more members, that if the node set has a single value, this value is returned, and if the node set is empty, a null value is returned. (The conditional behaviour in this case is intriguing: if there is a single value in the node set which is an array, that array will be returned and there is nothing to distinguish the return value from an array produced from a node set with two or more members.)

Reply:
At today's JSONPath WG interim meeting, the consensus was to allow only singular paths in comparisons and regex tests. This leave open whether we should allow non-singular paths for existence tests. My position is that it is more consistent and more memorable for users to require singular paths across the board in filter expressions than to have a different rule for existence tests.

@cabo may wish to jump in here.

Reply:
> I find this paragraph confusing.
> 
> "The slice, recursive descent and union selectors, which can evaluate to zero, one, or many items, are wrapped in a Json array. " -- I don't know what it means to wrap a selector in a JSON array. I presume it should say the node sets produced by these selectors are wrapped in JSON arrays, which sounds very much like automatically converting node sets into arrays.
> 
> "The object wildcard selector is wrapped in an array if the object has two or more members, otherwise a single value is returned." -- I presume you mean the node set produced by such a selector is wrapped (in a non-deterministic order?) in a JSON array if it has two or more members, that if the node set has a single value, this value is returned, and if the node set is empty, a null value is returned. (The conditional behaviour in this case is intriguing: if there is a single value in the node set which is an array, that array will be returned and there is nothing to distinguish the return value from an array produced from a node set with two or more members.)

Jayway's query function provides options to evaluate a JSONPath expression either as a result list or as a single value, the user can choose. Within filter expressions, though, an embedded JSONPath expression is always evaluated as a single value. For evaluating a JSONPath expression as a single value, the basic idea is that a selector that returns at most one value returns either that value or a JSON null (if none), and a selector that potentially can return zero or many values returns them as an array (rather than appending them to a result list). This isn't particularly well documented, but if you find my explanation unclear, you can refer to [JSONPath Reference](https://support.smartbear.com/readyapi/docs/testing/jsonpath-reference.html), the section on "Usage notes and considerations." Or if you're familiar with JMESPath, it's similar to how JMESPath evaluates expressions.

My main point though is as follows. 

Currently in the grammar, you have, for filter selectors
```
path         = rel-path / json-path
rel-path     = "@" *(S (dot-selector / index-selector))
```

If you allow a `json-path` in an expression, I think you, like Jayway and other implementations, have to answer the question: how do I evaluate the `json-path` as a single value? And if you answer that question (your answer may be different than Jayway's), why limit `rel-path`? Why not let expressions beginning with '@' be full JSONPath expressions in their own right, like they are in Jayway (and other implementations that invented their own expression language)? Why not allow JSONPath expressions to begin with either '$' or '@', as Jayway does, and Bettio  proposed? This is eminently doable, there is proven prior experience.

On the other hand, if you prefer the simplicity of limiting expressions beginning with `@` to 
```
"@" *(S (dot-selector / index-selector))
```
which is consistent with the implementations that dovetail on JavaScript, Python and PHP, why not interpret them in the same way as JavaScript, Python, and PHP do? As evaluating a variable '@' against identifier and index operators, which always return a single value? Where they are most emphatically _not_ JSONPath expressions, "singular" or otherwise?  Why not just drop `json-path` from the allowed elements in filters, so you don't have to tackle the problem of how to evaluate a JSONPath expression as a single value at all?

Reply:
> Why not just drop json-path from the allowed elements in filters, so you don't have to tackle the problem of how to evaluate a JSONPath expression as a single value at all?

I believe that's consistent with the direction we are taking.